sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1906,291 @@ def transform(self, X):
+        The inverse of the cumulative density function used to project the
+        data.
+
+    See also
+    --------
+    :class:`sklearn.preprocessing.StandardScaler` to perform standardization
+    that is faster, but less robust to outliers.
+    """
+
+    def __init__(self, n_quantiles=1000, subsample=int(1e5),
+                 random_state=None):
+        self.n_quantiles = n_quantiles
+        self.subsample = subsample
+        self.random_state = random_state
+
+    def _build_f(self):
+        """Build the transform functions."""
+        check_is_fitted(self, 'quantiles_')
+
+        self.f_transform_ = [
+            interp1d(quantiles_feat, self.references_,
+                     bounds_error=False,
+                     fill_value=(min(self.references_),
+                                 max(self.references_)))
+            for quantiles_feat in self.quantiles_.T]
+
+        self.f_inverse_transform_ = [
+            interp1d(self.references_, quantiles_feat,
+                     bounds_error=False,
+                     fill_value=(min(quantiles_feat),
+                                 max(quantiles_feat)))
+            for quantiles_feat in self.quantiles_.T]
+
+    def _dense_fit(self, X):
+        """Compute percentiles for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+        """
+        rng = check_random_state(self.random_state)
+
+        # subsample the matrix X if necessary
+        if self.subsample < X.shape[0]:
+            subsample_idx = rng.choice(X.shape[0], self.subsample,
+                                       replace=False)
+        else:
+            subsample_idx = range(X.shape[0])
+
+        self.references_ = np.linspace(0, 1, self.n_quantiles,
+                                       endpoint=True)
+        self.quantiles_ = np.percentile(X[subsample_idx, :],
+                                        self.references_ * 100, axis=0)
+
+    def _sparse_fit(self, X):
+        """Compute percentiles for sparse matrices.
+
+        Parameters
+        ----------
+        X : sparse matrix CSC, shape (n_samples, n_features)
+            The data used to scale along the features axis. The sparse matrix
+            needs to be semi-positive.
+        """
+        rng = check_random_state(self.random_state)
+
+        n_samples, n_feat = X.get_shape()
+        if self.subsample < n_samples:
+            subsample_idx = rng.choice(n_samples, self.subsample,
+                                       replace=False)
+            X_csr = X.tocsr()[subsample_idx]
+            X = X_csr.tocsc()
+
+        self.references_ = np.linspace(0, 1, self.n_quantiles,
+                                       endpoint=True)
+        # FIXME: it does not take into account the zero in the computation
+        self.quantiles_ = np.array([np.percentile(
```

Comment:
@ogrisel @tguillemot Here I am not really sure what should be the right way.
Assuming that the sparse matrix as a lot of zeros fo a given feature, it will have a bad influence on the normalisation, didn't it?
It could also be the case in the dense in fact. That was the reason of including a `quantile_range`.